In [106]:
# 1. before converting data into ridge regression, we have to do StandardScalar first

# #Create pipeline:

# 1. Standardize the predicitons
# 2. 





import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

In [107]:
data = {
    "Study_hour": [2,3,4,5,6,7,8,9],
    "aten" : [60,65,70,75,80,85,90,95],
    "ass_sc" : [50,55,58,65,70,74,80,88],
    "final_mark" : [42,48,54,61,67,73,81,89]
}

df = pd.DataFrame(data)

x = df.iloc[:,0:3]
y = df["final_mark"]

In [108]:
x

,Study_hour,aten,ass_sc
0,2,60,50
1,3,65,55
2,4,70,58
3,5,75,65
4,6,80,70
5,7,85,74
6,8,90,80
7,9,95,88


In [109]:
y

0    42
1    48
2    54
3    61
4    67
5    73
6    81
7    89
Name: final_mark, dtype: int64

In [110]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import train_test_split
scaler = StandardScaler()
X_scaled = scaler.fit_transform(x)             # ---- X is our independen varibale

In [111]:
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42
)


In [112]:
model = make_pipeline(
    StandardScaler(),
    Ridge(alpha=1.0)
)

model.fit(x_train, y_train)
y_pred = model.predict(x_test)
y_pred

array([49.09249889, 73.46107412])

In [113]:
from sklearn.linear_model import RidgeCV

model = RidgeCV(
    alphas= [0.01, 0.1, 1, 10,100],
    cv = 5
)

model.fit(x_train, y_train)

print("Best Alphas: ", model.alpha_)

Best Alphas:  0.01


c:\Users\bda\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\metrics\_regression.py:1295: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
c:\Users\bda\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\metrics\_regression.py:1295: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
c:\Users\bda\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\metrics\_regression.py:1295: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
c:\Users\bda\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\metrics\_regression.py:1295: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
c:\Users\bda\AppData\Local\Programs\Python\Python311\Lib\sit

In [114]:
model = make_pipeline(
    StandardScaler(),
    RidgeCV(alphas= [0.01, 0.1, 1, 10,100])
)

model.fit(x_train, y_train)

Ridge_model = model.named_steps["ridgecv"]

print("Best Alphas: ", Ridge_model.alpha_)

Best Alphas:  0.1


<h1>Lasso Implementation</h1>

In [115]:
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [116]:
data = {
    "SH" : [1,2,2,3,3,4,5,5,6,7,7,8,8,9,10],
    "ATT" : [55,60,6,62,70,72,75,80,78,82,85,88,90,92,95],
    "ASS_SCR" :[40,45,50,48,55,58,62,67,65,70,74,78,80,85,90],
    "STU_ID" : [101,102,103,104,105,106,107,108,109,110,111,112,113,114,115],
    "Fi_Ma" : [35,40,43,45,50,54,60,64,66,71,74,79,82,87,92]
}

df = pd.DataFrame(data)

df

,SH,ATT,ASS_SCR,STU_ID,Fi_Ma
0,1,55,40,101,35
1,2,60,45,102,40
2,2,6,50,103,43
3,3,62,48,104,45
4,3,70,55,105,50
5,4,72,58,106,54
6,5,75,62,107,60
7,5,80,67,108,64
8,6,78,65,109,66
9,7,82,70,110,71


In [117]:
x = df[
    [
        "SH",
        "ATT",
        "ASS_SCR",
        "STU_ID"
    ]
]
y  = df ["Fi_Ma"]
x

,SH,ATT,ASS_SCR,STU_ID
0,1,55,40,101
1,2,60,45,102
2,2,6,50,103
3,3,62,48,104
4,3,70,55,105
5,4,72,58,106
6,5,75,62,107
7,5,80,67,108
8,6,78,65,109
9,7,82,70,110


In [118]:
y

0     35
1     40
2     43
3     45
4     50
5     54
6     60
7     64
8     66
9     71
10    74
11    79
12    82
13    87
14    92
Name: Fi_Ma, dtype: int64

In [119]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42
)

model = make_pipeline(
    StandardScaler(),
    Lasso(alpha=0.1)
)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

lasso_model = model.named_steps["lasso"]

coefficient_table = pd.DataFrame({
    "Predicitor" : x.columns,
    "Lasso Coefficient" : lasso_model.coef_
})

print(coefficient_table)

  Predicitor  Lasso Coefficient
0         SH           4.074742
1        ATT           0.133936
2    ASS_SCR           6.849160
3     STU_ID           5.805057


In [120]:
mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R2 Score: ", r2)

MAE: 0.3104981980822856
MSE: 0.16697027837423398
RMSE: 0.408619968154071
R2 Score:  0.9995440738757985


<h3> Task on same data: 1 OLR   <br>2. Ridge    <br>3. Lasso    <br>4. Auto (RidgecV and LassoCv)<br>Compare which is best

In [121]:
#1. OLR

model = LinearRegression()
model.fit(x, y)
print("Intercept: ", model.intercept_)

Intercept:  -129.27024012943542


In [122]:
coefficient_table = pd.DataFrame({
    "Feature" : x.columns,
    "Coefficient" : model.coef_
})

print(coefficient_table)

   Feature  Coefficient
0       SH     1.541476
1      ATT     0.010962
2  ASS_SCR     0.475681
3   STU_ID     1.410990


In [123]:
mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)

print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R2 Score: ", r2)

MAE: 0.3104981980822856
MSE: 0.16697027837423398
RMSE: 0.408619968154071
R2 Score:  0.9995440738757985


<h3>Working On Ridge with same dataset

In [124]:
model = make_pipeline(
    StandardScaler(),
    Ridge(alpha=1.0)
)

model.fit(x_train, y_train)
y_pred = model.predict(x_test)
print(y_pred)

from sklearn.linear_model import RidgeCV

model = RidgeCV(
    alphas= [0.01, 0.1, 1, 10,100],
    cv = 5
)

model.fit(x_train, y_train)

print("Best Alphas: ", model.alpha_)

mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)

print("\nMAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R2 Score: ", r2)

coefficient_table = pd.DataFrame({
    "Feature" : x.columns,
    "Coefficient" : model.coef_
})

print("\n",coefficient_table)

[71.06386161 78.89733293 35.30556047]
Best Alphas:  1.0

MAE: 0.15736304795041178
MSE: 0.03599534377712476
RMSE: 0.18972438898867156
R2 Score:  0.999901711743327

    Feature  Coefficient
0       SH     1.132307
1      ATT     0.016355
2  ASS_SCR     0.603667
3   STU_ID     1.227705


In [128]:
from sklearn.linear_model import RidgeCV

model = RidgeCV(
    alphas= [0.01, 0.1,10,100],
    cv = 5,
)

model.fit(x_train, y_train)

print("Best Alphas: ", model.alpha_)

y_pred = model.predict(x_test)

mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)

print("\nMAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R2 Score: ", r2)


coefficient_table = pd.DataFrame({
    "Feature" : x.columns,
    "Coefficient" : model.coef_
})

print("\n",coefficient_table)

Best Alphas:  0.1

MAE: 0.38713605246593374
MSE: 0.24911688229010184
RMSE: 0.49911610101268206
R2 Score:  0.9993197657947176

    Feature  Coefficient
0       SH     1.489198
1      ATT     0.009709
2  ASS_SCR     0.504485
3   STU_ID     1.373094


<h3>Working on Same dataset using LassoCV

In [126]:
from sklearn.linear_model import LassoCV

model = make_pipeline(
    StandardScaler(),
    LassoCV(
        alphas=[0.001,0.1,1,10,100],
        cv = 5,
        max_iter=10000
    )
)
model.fit(x_train, y_train)

# Get the trained LassoCV model
lassoCv = model.named_steps['lassocv']

# Print the best alpha
print("Best alpha:", lassoCv.alpha_)

y_pred = model.predict(x_test)

mae = mean_absolute_error(y_test,y_pred)
mse = mean_squared_error(y_test,y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test,y_pred)

print("\nMAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("R2 Score: ", r2)


coefficient_table = pd.DataFrame({
    "Feature" : x.columns,
    "Coefficient" : lassoCv.coef_
})

print("\n",coefficient_table)

Best alpha: 0.1

MAE: 0.3104981980822856
MSE: 0.16697027837423398
RMSE: 0.408619968154071
R2 Score:  0.9995440738757985

    Feature  Coefficient
0       SH     4.074742
1      ATT     0.133936
2  ASS_SCR     6.849160
3   STU_ID     5.805057


<h3>Summary:

# Regression Model Results

This section presents the results obtained from five regression models:

- LassoCV
- RidgeCV
- Lasso
- Ridge
- OLS (Ordinary Least Squares)

The models were evaluated using MAE, MSE, RMSE, and R² Score.

## 1. LassoCV

**Best Alpha:** 0.1

**Performance Metrics**

| Metric | Value |
|---|---|
| MAE | 0.3104981980822856 |
| MSE | 0.16697027837423398 |
| RMSE | 0.408619968154071 |
| R² Score | 0.9995440738757985 |

**Feature Coefficients**

| Feature | Coefficient |
|---|---|
| SH | 4.074742 |
| ATT | 0.133936 |
| ASS_SCR | 6.849160 |
| STU_ID | 5.805057 |

## 2. RidgeCV

**Best Alpha:** 0.1

**Performance Metrics**

| Metric | Value |
|---|---|
| MAE | 0.38713605246593374 |
| MSE | 0.24911688229010184 |
| RMSE | 0.49911610101268206 |
| R² Score | 0.9993197657947176 |

**Feature Coefficients**

| Feature | Coefficient |
|---|---|
| SH | 1.489198 |
| ATT | 0.009709 |
| ASS_SCR | 0.504485 |
| STU_ID | 1.373094 |

## 3. Lasso Regression

**Performance Metrics**

| Metric | Value |
|---|---|
| MAE | 0.3104981980822856 |
| MSE | 0.16697027837423398 |
| RMSE | 0.408619968154071 |
| R² Score | 0.9995440738757985 |

**Feature Coefficients**

| Predictor | Lasso Coefficient |
|---|---|
| SH | 4.074742 |
| ATT | 0.133936 |
| ASS_SCR | 6.849160 |
| STU_ID | 5.805057 |

## 4. Ridge Regression

**Predictions**

```
[71.06386161 78.89733293 35.30556047]
```

**Best Alpha:** 1.0

**Performance Metrics**

| Metric | Value |
|---|---|
| MAE | 0.15736304795041178 |
| MSE | 0.03599534377712476 |
| RMSE | 0.18972438898867156 |
| R² Score | 0.999901711743327 |

**Feature Coefficients**

| Feature | Coefficient |
|---|---|
| SH | 1.132307 |
| ATT | 0.016355 |
| ASS_SCR | 0.603667 |
| STU_ID | 1.227705 |

## 5. OLS Regression

**Performance Metrics**

| Metric | Value |
|---|---|
| MAE | 0.3104981980822856 |
| MSE | 0.16697027837423398 |
| RMSE | 0.408619968154071 |
| R² Score | 0.9995440738757985 |

**Feature Coefficients**

| Feature | Coefficient |
|---|---|
| SH | 1.541476 |
| ATT | 0.010962 |
| ASS_SCR | 0.475681 |
| STU_ID | 1.410990 |

## 6. Overall Model Comparison

| Model | Best Alpha | MAE | MSE | RMSE | R² Score |
|---|---|---|---|---|---|
| LassoCV | 0.1 | 0.310498 | 0.166970 | 0.408620 | 0.999544 |
| RidgeCV | 0.1 | 0.387136 | 0.249117 | 0.499116 | 0.999320 |
| Lasso | — | 0.310498 | 0.166970 | 0.408620 | 0.999544 |
| Ridge | 1.0 | 0.157363 | 0.035995 | 0.189724 | 0.999902 |
| OLS | — | 0.310498 | 0.166970 | 0.408620 | 0.999544 |

## 7. Feature Coefficient Comparison

| Feature | LassoCV | RidgeCV | Lasso | Ridge | OLS |
|---|---|---|---|---|---|
| SH | 4.074742 | 1.489198 | 4.074742 | 1.132307 | 1.541476 |
| ATT | 0.133936 | 0.009709 | 0.133936 | 0.016355 | 0.010962 |
| ASS_SCR | 6.849160 | 0.504485 | 6.849160 | 0.603667 | 0.475681 |
| STU_ID | 5.805057 | 1.373094 | 5.805057 | 1.227705 | 1.410990 |

## 8. Model Performance Analysis

Based on the evaluation metrics, Ridge Regression achieved the best performance among the five models.

**Ridge Regression Results**
- Best Alpha: 1.0
- MAE: 0.15736304795041178
- MSE: 0.03599534377712476
- RMSE: 0.18972438898867156
- R² Score: 0.999901711743327

Ridge Regression achieved the:
- Lowest MAE
- Lowest MSE
- Lowest RMSE
- Highest R² Score

Therefore, Ridge Regression with alpha = 1.0 is the best-performing model based on the evaluated metrics.

## 9. Model Ranking

Based on the R² Score and error metrics:

| Rank | Model | R² Score |
|---|---|---|
| 1 | Ridge | 0.999901711743327 |
| 2 | LassoCV | 0.9995440738757985 |
| 2 | Lasso | 0.9995440738757985 |
| 2 | OLS | 0.9995440738757985 |
| 3 | RidgeCV | 0.9993197657947176 |

## 10. Conclusion

Among all the evaluated models, Ridge Regression performed the best.

The Ridge model with alpha = 1.0 achieved:

- MAE = 0.157363
- MSE = 0.035995
- RMSE = 0.189724
- R² Score = 0.999902

The R² score of approximately 99.99% indicates that the model explains almost all of the variation in the target variable on the evaluated dataset.

LassoCV, Lasso, and OLS produced identical evaluation metrics in this experiment, while RidgeCV achieved a slightly lower R² score.

Overall, Ridge Regression is the preferred model for this dataset based on the current evaluation results.
